Fix non-parallel make under depfile jobserver

Jordan474 opened the pull request:
Fix #32861 

When a package asks for non-parallel make, we need to force `make -j1` because just doing `make` will run in parallel under jobserver (e.g. `spack env depfile`).

We now always add `-j1` when asked for a non-parallel execution (even if there is no jobserver).

And each `MakeExecutable` can now ask for jobserver support or not. For example: the default `ninja` does not support jobserver so spack applies the default `-j`, but `ninja@kitware` or `ninja-fortran` does, so spack doesn't add `-j`.

Tips: you can run `SPACK_INSTALL_FLAGS=-j1 make -f spack-env-depfile.make -j8` to avoid massive job-spawning because of build tools that don't support jobserver (ninja).

Commit:
Fix non-parallel make under depfile jobserver

When a package ask for non-parallel make, we need to force `make -j1`
because just doing `make` will run in parallel under depfile jobserver.

Also, MakeExecutable can now say if it supports jobserver or not. This
is useful to limit ninja under jobserver, but still support it for
Kitware's patched ninja.

haampie commented on lib/spack/spack/build_environment.py:
```diff
@@ -150,14 +150,19 @@ def __call__(self, *args, **kwargs):
         parallel = should_set_parallel_jobs(jobserver_support=True) and kwargs.pop(
             "parallel", self.jobs > 1
         )
+        if not parallel:
+            # Force -j1 to disable parallel even under jobserver
+            jobs = 1
```
this branch is now followed when job server support is detected in general, which is incorrect.

Jordan474 replied:
Yes, the first version did, but then the test failed, and I did a mistake in the new version. I'll mark this PR as draft until I find a proper fix.

Jordan474 commented:
It took me a few tries. There is few parameters to think about:
- `spack install -j`
- `make(parallel=false)`
- `SPACK_NO_PARALLEL_MAKE`
- jobserver or not

I changed `should_set_parallel_jobs` because I think the semantic was confusing: when non-parallel but with a jobserver we should still "set parallel jobs" to 1. So I changed it a more explicit `jobserver_enabled()`. See d3a0ac1c0ab5e04cf0636362b0861e23d9e3018b.

I think explicit `jobserver_enabled()` handling in `package.py` is fine because you need to know if the tool behind actually support jobserver.

There may be a more generic refactor to find but would probably be too disruptive (`determine_number_of_jobs(support_jobserver=True) -> Optional[int]` ? what about `make_jobs`, `package.parallel` ?).

Jordan474 commented:
> There may be a more generic refactor to find but would probably be too disruptive

Note on this possible refactor: maybe the default `make_jobs` etc. could be lowered for tools that don't explicitly support jobserver. One can already emulate this via `SPACK_INSTALL_FLAGS=-j4 make -f depfile.make -j16`, as a security to avoid worst case scenario of 16 * 16 tasks running because of non-jobserver-compatible builds.

haampie commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -310,7 +310,7 @@ def bootstrap_args(self):
             args.append("--prefix={0}".format(self.prefix))
 
-            if spack.build_environment.should_set_parallel_jobs(jobserver_support=True):
+            if not spack.build_environment.jobserver_enabled():
```
This was actually the reason there was a `should_set_parallel_jobs` function: the idea was to have it return true if and only if `-j` should be passed explicitly.

With this change, cmake does not respect `SPACK_NO_PARALLEL_MAKE` anymore

Jordan474 replied:
I consciously dropped the support of `SPACK_NO_PARALLEL_MAKE` here because of the `-j`/not-`-j` confusion, and it didn't seem supported before the introduction of `should_set_parallel_jobs`.

I'll work on reintroducing it.

haampie reviewed (changes requested):
Can you make the cmake build respect `SPACK_NO_PARALLEL_MAKE` again?

If it simplifies the code, I'm also happy with we always invoke `make -j1` when `parallel=False` instead of `make`, so you'd get something along the lines of:

```python
def get_effective_jobs(jobs, parallel=True):
    """Returns the effective number of jobs to be specified on the command line when invoking make

    Returns:
        int or None: an integer to be passed as a -j <n> flag, or None when no -j flag should be passed
    """
    if not parallel or env_flag(SPACK_NO_PARALLEL_MAKE):
        return 1
    elif jobserver_enabled():
        return None
    return jobs
```

Jordan474 commented:
Updated:

Fix `SPACK_NO_PARALLEL_MAKE` regression: now behaves like before this MR

Always force `-j1` when non-parallel, even when there is no jobserver

Each `MakeExecutable` can now ask to disable jobserver support (this fixes a comment about ninja by @trws).  Jobserver support is disabled for `ninja`, but re-enabled for `ninja-fortran` Kitware's fork, and could be re-enabled for `ninja` if https://github.com/ninja-build/ninja/pull/1140 is merged.

(Updated description)

haampie commented on var/spack/repos/builtin/packages/ninja/package.py:
```diff
@@ -79,4 +79,5 @@ def setup_dependent_package(self, module, dspec):
         module.ninja = MakeExecutable(
             which_string(name, path=[self.spec.prefix.bin], required=True),
             determine_number_of_jobs(parallel=self.parallel),
+            supports_jobserver=False,
```
```suggestion
            supports_jobserver=self.spec.version == ver('kitware'),
```

haampie commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -307,11 +307,20 @@ def bootstrap_args(self):
+            args.append("-GNinja")
+            self.generator = ninja
+
         if not sys.platform == "win32":
```
```suggestion
        else:
```

haampie reviewed (changes requested):
Looks good, we have the kitware fork here, so if you incorporate my suggestion I think this is good to go.

If you're a `depfile` user yourself, use

```yaml
spack:
  specs: [...]
  packages:
    ninja:
      require: '@kitware'
```

Jordan474 committed:
Support jobserver for ninja@kitware

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

haampie commented:
@spackbot fix style

trws commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -307,11 +307,20 @@ def bootstrap_args(self):
 
-            if spack.build_environment.should_set_parallel_jobs(jobserver_support=True):
-                args.append("--parallel={0}".format(make_jobs))
+            jobs = get_effective_jobs(
```
Missing import

Jordan474 replied:
Thanks, fixed.

trws reviewed (changes requested):
As soon as this passes I'm good, comment on the current breakage above.

haampie committed:
Merge branch 'develop' into fix-non-parallel-under-jobserver

Jordan474 committed:
Merge branch 'develop' into fix-non-parallel-under-jobserver

Jordan474 committed:
Merge branch 'develop' into fix-non-parallel-under-jobserver

Jordan474 committed:
Merge branch 'develop' into fix-non-parallel-under-jobserver

Jordan474 commented:
The gitlab pipeline [keeps failing](https://gitlab.spack.io/spack/spack/-/jobs/4170628#L174), but I don't understand how this could come from my changes ?

tgamblin commented:
The issues this is getting don't look like they're from your changes.  We're working on this!  Sorry for the false alarms!

haampie commented:
@spackbot run pipeline

haampie commented:
Can you just hit [merge] @tgamblin?

tgamblin commented:
Done
